das/aggregator.go

```diff
@@ -0,0 +1,218 @@
+			copy(blobHash[:], crypto.Keccak256(blob))
+			if blobHash == requestedCert.DataHash {
+				blobChan <- blob
+			} else {
+				errorChan <- fmt.Errorf("DAS (mask %X) returned data that doesn't match requested hash!", d.signerMask)
+			}
+		}(subCtx, d)
+	}
+
+	errorCount := 0
+	var errorCollection []error
+	for errorCount < len(servicesThatSignedCert) {
+		select {
+		case blob := <-blobChan:
+			return blob, nil
+		case err = <-errorChan:
+			errorCollection = append(errorCollection, err)
+			log.Warn("Couldn't retrieve message from DAS", "err", err)
+			errorCount++
+		case <-ctx.Done():
+			break
+		}
+	}
+
+	return nil, fmt.Errorf("Data wasn't able to be retrieved from any DAS: %v", errorCollection)
+}
+
+type storeResponse struct {
+	cert *arbstate.DataAvailabilityCertificate
+	err  error
+
+	details serviceDetails
+}
+
+// Store calls Store on each backend DAS in parallel and collects responses; if
+// there were enough (at least K) responses then it aggregates the signatures and
+// signerMasks from each DAS together to put in the DataAvailabilityCertificate
+// that it returns; if it gets enough errors that K successes is impossible, then
+// it stops early and returns an error, and if it gets not enough successful
+// responses by the time its context is canceled (eg via DeadlineWrapper) then it
+// also returns an error.
+func (a *Aggregator) Store(ctx context.Context, message []byte, timeout uint64) (*arbstate.DataAvailabilityCertificate, error) {
+	var aggSignersMask uint64
+	var pubKeys []blsSignatures.PublicKey
+	var sigs []blsSignatures.Signature
+	var aggCert arbstate.DataAvailabilityCertificate
+
+	var initialStoreSucceeded bool
+	storeFailures := 0
+	if timeout == CALLEE_PICKS_TIMEOUT {
+		timeout = uint64(time.Now().Add(a.config.retentionPeriod).Unix())
+	}
+
+	responses := make(chan storeResponse)
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for _, d := range a.services {
+		go func(ctx context.Context, d serviceDetails) {
+			cert, err := d.service.Store(ctx, message, timeout)
+			responses <- storeResponse{cert, err, d}
+		}(subCtx, d)
+	}
+
+	successfullyStoredCount := 0
+	for i := 0; i < len(a.services); i++ {
+		select {
+		case <-ctx.Done():
+			if successfullyStoredCount < a.requiredServicesForStore {
+				return nil, errors.New("Terminated das.Aggregator.Store() with requests outstanding before enough responses received")
+			} else {
+				break
+			}
+		case r := <-responses:
+			if r.err != nil {
+				storeFailures++
+				log.Warn("Failed to store message to DAS", "err", r.err)
+				if storeFailures <= a.maxAllowedServiceStoreFailures {
+					continue
+				} else {
+					return nil, fmt.Errorf("Aggregator failed to store message to at least %d out of %d DASes (assuming %d are honest)", a.requiredServicesForStore, len(a.services), a.config.assumedHonest)
+				}
+			}
+			verified, err := blsSignatures.VerifySignature(r.cert.Sig, serializeSignableFields(*r.cert), r.details.pubKey)
```

Comment:
We shouldn't trust the contents of `r.cert` here, because it came from a DAS that might be malicious. Se we should check the signature against our own version of the signable cert fields, and our own version of the public key.

Reply:
Actually it's probably better to not pass any untrusted data on the `responses` channel. That way we can't accidentally use the untrusted data.

Reply:
I changed it to pass only the signature from the cert on the channel. The other fields are validated against the expected values (we had discussed that they also could be ignored, but for now I chose to validate them). This made it easy to simplify the `Store` method by moving more checking logic into the goroutines for each backend, so the response gathering is quite simple now.